lib/puppet/provider/nssdb/certutil.rb

```diff
@@ -0,0 +1,50 @@
+require 'fileutils'
+
+Puppet::Type.type(:nssdb).provide(:certutil) do
+  commands :certutil => 'certutil'
+
+  def create
+    destroy_nssdb
+    create_nssdb_dir
+    generate_nssdb
+  end
+
+  def destroy
+    destroy_nssdb
+  end
+
+  def exists?
+    nssdb_exists?
+  end
+
+  private
+
+  def create_nssdb_dir
+    FileUtils.mkdir_p(resource[:nssdb_dir])
+  end
+
+  def generate_nssdb
+    certutil(
+      '-N',
+      '-d', resource[:nssdb_dir],
+      '-f', resource[:password_file]
+    )
+  rescue Puppet::ExecutionFailure => e
+    raise Puppet::Error.new("Failed to generate new NSS database at #{resource[:nssdb_dir]} with password file #{resource[:password_file]}: #{e}", e)
+  end
+
+  def destroy_nssdb
+    FileUtils.rm_rf(resource[:nssdb_dir], secure: true)
+  end
+
+  def nssdb_exists?
+    certutil(
+      '-L',
+      '-d', resource[:nssdb_dir],
+      '-f', resource[:password_file]
+    )
```

Comment:
I led you astray here slightly, turns out that to view a NSS DB you do not need the password, and its not even an option for `-L`.

```
-L              List all certs, or print out a single named cert (or a subset)
   -h token-name     Name of token to search ("all" for all tokens)
   -n cert-name      Pretty print named cert (list all if unspecified)
   --email email-address 
                     Pretty print cert with email address (list all if unspecified)
   -d certdir        Cert database directory (default is ~/.netscape)
   -P dbprefix       Cert & Key database prefix
   -X                force the database to open R/W
   -r                For single cert, print binary DER encoding
   -a                For single cert, print ASCII encoding (RFC1113)
   --dump-ext-val OID 
                     For single cert, print binary DER encoding of extension OID
```

However, what does work is to try and list all private keys:

```
# certutil -K -d /etc/pki/katello/nssdb/ -f /etc/pki/katello/nssdb/nss_db_password-file 
certutil: Checking token "NSS Certificate DB" in slot "NSS User Private Key and Certificate Services"
< 0> rsa      f2e5cc6d95dcef5d1edfeb34bc0818c7fc1f2cc1   broker
```

And if bad password:

```
# certutil -K -d /etc/pki/katello/nssdb/ -f /etc/pki/katello/nss_db_password-file 
certutil: Checking token "NSS Certificate DB" in slot "NSS User Private Key and Certificate Services"
Incorrect password/PIN entered.
certutil: could not authenticate to token NSS Certificate DB.: SEC_ERROR_BAD_PASSWORD: The security password entered is incorrect.
```

I found this test running an upgrade pipeline to ensure this would work as we have theorized.
```
```

Reply:
Updated to use `-K`

If you run the pipeline again, please let me know where I could observe. TY!